llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp

```diff
@@ -2797,6 +2784,50 @@ static bool insertSinCosCall(IRBuilderBase &B, Function *OrigCallee, Value *Arg,
   return true;
 }
 
+static Value *optimizeSymmetricCall(CallInst *CI, bool IsEven,
+                                    IRBuilderBase &B) {
+  Value *X;
+  if (match(CI->getArgOperand(0), m_OneUse(m_FNeg(m_Value(X))))) {
+    IRBuilderBase::FastMathFlagGuard Guard(B);
+    B.setFastMathFlags(CI->getFastMathFlags());
+
+    auto *CallInst = copyFlags(*CI, B.CreateCall(CI->getCalledFunction(), {X}));
+    if (IsEven) {
+      // Even function: f(-x) = f(x)
+      return CallInst;
+    }
+    // Odd function: f(-x) = -f(x)
+    return B.CreateFNeg(CallInst);
```

Comment:
On a side note, I wonder if commuting the sign with an odd function is an improvement in practice. For even functions, we're clearly getting rid of an instruction, which is almost always a win. But for odd functions, it's just moving the instruction, and it's not offhand clear to me if peephole opportunities tend to be better with it being before or after the libm function.

Reply:
I guess one option is to look only for cases like `-f(-x)` where `f` is odd, so it can be replaced with just `f(x)`.